rfcs/134-write-behavior-consistency.md

```diff
@@ -0,0 +1,227 @@
+- Author: Xuanwo <github@xuanwo.io>
+- Start Date: 2021-07-04
+- RFC PR: [beyondstorage/specs#134](https://github.com/beyondstorage/specs/pull/134)
+- Tracking Issue: [beyondstorage/go-storage#0](https://github.com/beyondstorage/go-storage/issues/0)
+
+# GSP-134: Write Behavior Consistency
+
+Previous Discussions:
+
+- [Add detailed behavior about move/copy](https://github.com/beyondstorage/specs/issues/130)
+- [Make write Idempotent](https://github.com/beyondstorage/specs/issues/96)
+- [copy/move doesn't need to support dst path that has an existing path](https://github.com/beyondstorage/go-integration-test/issues/29)
+
+## Background
+
+Behavior consistency is extremely critical to us, and write related operations is the most important part.
+
+Write related operations including following operations:
+
+- `Write`
+- `CompleteMultipart`
+- `CombineBlock`
+- `CommitAppend`
+- `Copy`
+- `Move`
+
+All of them will create a new object in our storage service. But our services could have different behavior while the destination path already exists. I will list a few representative service behaviors for analysis.
+
+### fs
+
+For `Write`:
+
+`fs` could return `EEXIST` to indicate `File exists`. Users can set open mode `Create` to create a new file here without meeting `EEXIST` error.
+
+For `Copy`:
+
+`fs` doesn't have native copy support, we could implement it via reading from src and write into dst. For the writing part, most userland tools will use `Create` open mode to ignore `EEXIST` error.
+
+So `Copy` will not return the `EEXIST` error.
+
+For `Move`:
+
+`fs` will ignore the `EEXIST` error and always replace the file. On the same filesystem, it's an atomic operation, otherwise, it should be implemented as `copy-delete`.
+
+So `Move` will not return the `EEXIST` error.
+
+### s3
+
+Most s3 alike object storage services will not return object exist errors.
+
+- For object storage services that don't have an object version, the last write will always succeed.
+- For object storage services that have object versions, every write will create a new object version.
+
+### ipfs
+
+For `Write`:
+
+`ipfs` supports `Create` like `fs`, with `Create`, `Write` will not return `FileExists`.
+
+```go
+func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
+   err = s.ipfs.FilesWrite(
+      ctx, s.getAbsPath(path), r,
+      ipfs.FilesWrite.Create(true),
+      ipfs.FilesWrite.Parents(true),
+   )
+   if err != nil {
+      return 0, err
+   }
+   return size, nil
+}
+```
+
+For `Copy` and `Move`:
+
+`ipfs` have native support for `Copy` and `Move`, and will return `FileExists` if dst already exists.
```

Comment:
If the dst exists, the `copy` operation will return `FileExists`, while the `move` operation will overwrite.

Reply:
One thing to note is that if `src` matches IPFS-path, IPFS will actually add a reference to the root node of the file dag.

```
It can also be used to copy files within MFS, but in the case when an
IPFS-path matches an existing MFS path, the IPFS path wins.
```

Reply:
Does IPFS-path here mean an IPFS hash like `QmPK1s3pNYLi9ERiq3BDxKa4XosgWwFRQUydHUtz4YgpqB`?

I have no idea about it so far, maybe we can add an issue in ipfs repo or leave comments for `write`?

Reply:
I did not find clear documentation defining IPFS-path.

In the library [ipfs/go-path](https://github.com/ipfs/go-path) there is this comment:

https://github.com/ipfs/go-path/blob/a680d16b63280f753bdcf926a58868b46baeb25f/path.go#L12

```
// A Path represents an ipfs content path:
//   * /<cid>/path/to/file
//   * /ipfs/<cid>
//   * /ipns/<cid>/path/to/folder
//   * etc
type Path string
```

However, after my testing, the `src` of the `cp` operation only supports the following cases:

- `/ipfs/<CID>`, e.g., `/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX`
- `/ipfs/<CID>/path/to/file`, e.g., `/ipfs/QmT5NvUtoM5nWFfrQdVrFtvGfKFmG7AHE8P34isapyhCxX/wiki/Mars.html`
- files that already exist in MFS

I found supporting evidence in the code: 
- https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/core/commands/files.go#L387
- https://github.com/ipfs/go-ipfs/blob/ef866a1400b3b2861e5e8b6cc9edc8633b890a0a/core/commands/files.go#L408

```
node, err := getNodeFromPath(req.Context, nd, api, src)
if err != nil {
	return fmt.Errorf("cp: cannot get node from path %s: %s", src, err)
}

func getNodeFromPath(ctx context.Context, node *core.IpfsNode, api iface.CoreAPI, p string) (ipld.Node, error) {
	switch {
	case strings.HasPrefix(p, "/ipfs/"):
		return api.ResolveNode(ctx, path.New(p))
	default:
		fsn, err := mfs.Lookup(node.FilesRoot, p)
		if err != nil {
			return nil, err
		}

		return fsn.GetNode()
	}
}
```